Fixed issue 192

marchy opened the pull request:
Fixed issue 192 (https://github.com/BoltsFramework/Bolts-iOS/issues/192), allowing projects that include frameworks which themselves include Bolts to work without running into “Include of non-modular header inside framework module” compilation errors.

The problem here was that there were dependencies included directly in the .h file instead of in the .m files (see here: http://stackoverflow.com/questions/28552500/xcode6-receiving-error-include-of-non-modular-header-inside-framework-module).

Instead these were moved to use the proper @class and @protocol definitions in the headers and only including the actual file in the .m file. This removes the circular dependency issues so that the Objective-C compiler can properly make everything work in projects that include sub-dependencies (ie: include a framework which it itself includes the Bolts library)

nlutsenko commented:
Let's have a broader discussion in #192.
The inclusion of dependencies in the framework header files is supported use case and used everywhere, say in Foundation/UIKit frameworks.

marchy commented:
The fix above solves the issues. This was only problematic in 3 files so it's a fairly small change. Summary of changes:
- Changed from use of imports to class forwarding
- Moved the two protocols that were defined in the headers into their own respective files (BFAppLinkNavigationType.h and BFAppLinkReturnToRefererViewDelegate.h)

This does solve the issue and is either way a best-practice in general, improving compilation speed if nothing else. Please have a look and accept the pull request if this works. Clearly the issue is happening for some people and it's much better to solve these issues rather than depend on workarounds such as CLANG_ALLOW_NON_MODULAR_INCLUDES_IN_FRAMEWORK_MODULES = YES to get around them (http://stackoverflow.com/questions/24103169/swift-compiler-error-non-modular-header-inside-framework-module)

nlutsenko commented on Bolts/Common/Bolts.h:
```diff
@@ -8,7 +8,6 @@
  *
  */
 
-#import <Bolts/BoltsVersion.h>
```
This is still important to have.

nlutsenko commented on Bolts/iOS/BFAppLinkNavigation.h:
```diff
@@ -9,20 +9,8 @@
-/*!
- The result of calling navigate on a BFAppLinkNavigation
- */
-typedef NS_ENUM(NSInteger, BFAppLinkNavigationType) {
```
This doesn't seem necessary, as it's only relevant for usage together with BFAppLinkNavigation.

nlutsenko commented on Bolts/iOS/BFAppLinkReturnToRefererView.h:
```diff
@@ -23,25 +25,6 @@ typedef NS_ENUM(NSUInteger, BFIncludeStatusBarInSize) {
- Protocol that a class can implement in order to be notified when the user has navigated back
- to the referer of an App Link.
- */
-@protocol BFAppLinkReturnToRefererViewDelegate <NSObject>
```
This doesn't seem necessary, as it's only relevant for usage together with BFAppLinkReturnToRefererView.

nlutsenko commented:
Few nits here and there... I am sorry, but this still seems not necessary per discussion outlined in #192.

Class forwarding is a good approach, I absolutely agree though, but we are not using it in a lot of places, so I am somewhat sceptical about actual compilation impact.

nlutsenko commented:
Hey @marchy, I am going to close out this pull request as it looks abandoned.
Feel free to comment or rebase & reopen it.
